### PR TITLE
[Core] Undeprecating bool serialization

### LINKED ISSUE #9381: Verbose compilation warning in FluidDynamicsApplication
**Description**
More than a bug report, this is a warning report. Recently I fixed most issues with the fluid application (#9367) and one of the remaining ones stood out out because of its complexity. The output is quite involved:

```
[ 92%] Linking CXX shared module Kratos.cpython-38-x86_64-linux-gnu.so
In file included from /home/jbravo/Work/Kratos/kratos/includes/kratos_components.h:27,
                 from /home/jbravo/Work/Kratos/kratos/includes/kratos_flags.h:27,
                 from /home/jbravo/Work/Kratos/kratos/processes/process.h:25,
                 from /home/jbravo/Work/Kratos/applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.h:21,
                 from /home/jbravo/Work/Kratos/applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.cpp:16:
/home/jbravo/Work/Kratos/kratos/includes/data_communicator.h: In instantiation of ‘TObject Kratos::DataCommunicator::SendRecvImpl(const TObject&, int, int, int, int) const [with TObject = std::vector<bool>]’:
/home/jbravo/Work/Kratos/kratos/includes/data_communicator.h:525:34:   required from ‘TObject Kratos::DataCommunicator::SendRecv(const TObject&, int, int) const [with TObject = std::vector<bool>]’
/home/jbravo/Work/Kratos/kratos/utilities/pointer_communicator.h:238:63:   required from ‘void Kratos::GlobalPointerCommunicator<TPointerDataType>::Update(TFunctorType&, Kratos::GlobalPointersUnorderedMap<TPointerDataType, typename Kratos::ResultsProxy<TPointerDataType, TFunctorType>::TSendType>&) [with TFunctorType = Kratos::DistanceSmoothingProcess<2, Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::LinearSolver<Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > > >::Execute()::<lambda(Kratos::GlobalPointer<Kratos::Node<3> >&)>; TPointerDataType = Kratos::Node<3>; typename Kratos::ResultsProxy<TPointerDataType, TFunctorType>::TSendType = bool]’
/home/jbravo/Work/Kratos/kratos/utilities/pointer_communicator.h:214:19:   required from ‘Kratos::ResultsProxy<TPointerDataType, TFunctorType> Kratos::GlobalPointerCommunicator<TPointerDataType>::Apply(TFunctorType&&) [with TFunctorType = Kratos::DistanceSmoothingProcess<2, Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::LinearSolver<Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > > >::Execute()::<lambda(Kratos::GlobalPointer<Kratos::Node<3> >&)>; TPointerDataType = Kratos::Node<3>]’
/home/jbravo/Work/Kratos/applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.h:209:48:   required from ‘void Kratos::DistanceSmoothingProcess<TDim, TSparseSpace, TDenseSpace, TLinearSolver>::Execute() [with unsigned int TDim = 2; TSparseSpace = Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >; TDenseSpace = Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >; TLinearSolver = Kratos::LinearSolver<Kratos::UblasSpace<double, boost::numeric::ublas::compressed_matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > >]’
/home/jbravo/Work/Kratos/applications/FluidDynamicsApplication/custom_processes/distance_smoothing_process.cpp:25:16:   required from here
/home/jbravo/Work/Kratos/kratos/includes/data_communicator.h:880:40: warning: ‘void Kratos::DataCommunicator::CheckSerializationForSimpleType(const T&, Kratos::DataCommunicator::TypeFromBool<false>) const [with T = std::vector<bool>]’ is deprecated [-Wdeprecated-declarations]
  880 |         CheckSerializationForSimpleType(rSendObject, TypeFromBool<serialization_is_required<TObject>::value>());
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jbravo/Work/Kratos/kratos/includes/data_communicator.h:314:10: note: declared here
  314 |     void CheckSerializationForSimpleType(const T& rSerializedType, TypeFromBool<false>) const {}
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
The warning that is failing to print is:
```
Calling serialization-based communication for a simple type. Please implement direct communication support for this type.
```

Anyways any input from someone who knows about this stuff will be much appreciated.

**Scope**
Where in Kratos is this bug happening. 
- Compiling FluidDynamicsApplication with `g++`

**To Reproduce**
I managed to reproduced in two ways:
- Locally with `g++-11`.
- In the centos CI run as well, although not in the `gcc` one strangely enough.

**Expected behavior**
No warning

**Environment**
*Scenario 1*
- OS: Ubuntu 20.04.3 LTS
- Branch: master
- Compiler: `g++-11`

*Scenario 2*
Centos' CI run.

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
**📝 Description**
This PR is created to allow the communication of boolean values, since serialization is deprecated.

This PR originates from issue #9381.

**🆕 Changelog**
- Removed warning when attempting to use vectors of bools.

closes #9381

### CODE REVIEW

File: kratos/mpi/sources/mpi_data_communicator.cpp [hunk 1]
```diff
@@ -470,6 +471,8 @@ void MPIDataCommunicator::CheckMPIErrorCode(const int ierr, const std::string& M
 
 // Protected interface reimplementing DataCommunicator calls
 
+KRATOS_MPI_DATA_COMMUNICATOR_DEFINE_IMPLEMENTATION_FOR_TYPE(bool)
+KRATOS_MPI_DATA_COMMUNICATOR_DEFINE_IMPLEMENTATION_FOR_TYPE(bool)
```
Comment (philbucher) [inline kratos/mpi/sources/mpi_data_communicator.cpp]:
```suggestion
```

**Comment by EduardGomezEscandell** [comment]
We have an issue here:
https://github.com/KratosMultiphysics/Kratos/blob/c92658422d41bd149915af9fe933bcd24cb31d93/kratos/mpi/includes/mpi_message.h#L120-L123
since `void std::vector<bool, _Alloc>::data()` is deleted.

I see various ways out:
- *Minimize communication:* Convert the vector to a vector of some fixed-width integer type and do some bitwise logic to pack/unpack.
- *Compromise:* Convert to a vector of integers (the smallest ones, maybe `std::uint_least8_t`) and communicate it with a regular integer interface.
- *Minimize complexity:* Pass the problem along to the user and have them cast the values to `ints` outside the communicator.

I'd personaly side for the third one, since this doesn't seem to be a common issue.

**Comment by philbucher** [comment]
can you try `&rValue[0]` instead?

**Comment by philbucher** [comment]
> Besides this, std::vector<bool> does not necessarily store its elements as a contiguous array so it wouldn't work anyways. I don't believe there is any way of accessing the underlying data for this specialization.

really? I though std::vector guarantees this. After some brief research it seems like this is true, good to know

In this case I would have the user deal with this, i.e. use int or sth else. And a `static_assert` should be added to avoid this kind of problems in the future

**Comment by EduardGomezEscandell** [comment]
>  I though std::vector guarantees this

Yeah `vector<bool>` seems to break all usual rules.

> In this case I would have the user deal with this, i.e. use int or sth else. And a static_assert should be added to avoid this kind of problems in the future

At the moment we get a warning because it tries to use the deprecated serialization infrastructure, I suppose I can add a specific  warning for `vector<bool>`.

File: kratos/includes/data_communicator.h [hunk 2]
```diff
@@ -313,6 +313,10 @@ class KRATOS_API(KRATOS_CORE) DataCommunicator
     KRATOS_DEPRECATED_MESSAGE("Calling serialization-based communication for a simple type. Please implement direct communication support for this type.")
     void CheckSerializationForSimpleType(const T& rSerializedType, TypeFromBool<false>) const {}
 
+    template<bool value>
+    KRATOS_DEPRECATED_MESSAGE("Calling serialization-based communication for type bool. Communicating bools is generally problematic due to the nature of std::vector<bool>. It is recommended to cast them to ints for communication purposes.")
+    void CheckSerializationForSimpleType(const std::vector<bool>& rSerializedType, TypeFromBool<value>) const {}
```
Comment (philbucher) [inline kratos/includes/data_communicator.h]: hm so now it uses serialization for bool?
I think we can live with this, we only need to take out the message for bool (as bools can only be exchanges with serialization) then we can keep the support for bool and no longer have the warning

Comment (EduardGomezEscandell) [reply kratos/includes/data_communicator.h]: Yeah it falls back to serialization by default.

Ok, so I'll remove the warning for bool.

Comment (philbucher) [reply kratos/includes/data_communicator.h]: yes that makes sense to me

**Review by philbucher — APPROVED** [review]
cool

so now the warning is gone right?

**Comment by EduardGomezEscandell** [comment]
> cool
> 
> so now the warning is gone right?

Yes :smile:

**Pull request merged**